packages/transactions/src/codecs/transaction-codec.ts

```diff
@@ -10,3 +29,80 @@ export function getNewTransactionEncoder(): VariableSizeEncoder<NewTransaction>
     ]);
 }
+
+export function getNewTransactionDecoder(): VariableSizeDecoder<NewTransaction> {
+    return transformDecoder(
+        getStructDecoder([
+            ['signatures', getArrayDecoder(fixDecoderSize(getBytesDecoder(), 64), { size: getShortU16Decoder() })],
+            ['messageBytes', getBytesDecoder()],
+        ]),
+        decodePartiallyDecodedTransaction,
+    );
+}
+
+export function getNewTransactionCodec(): VariableSizeCodec<NewTransaction> {
+    return combineCodec(getNewTransactionEncoder(), getNewTransactionDecoder());
+}
+
+type PartiallyDecodedTransaction = {
+    messageBytes: ReadonlyUint8Array;
+    signatures: ReadonlyUint8Array[];
+};
+
+function decodePartiallyDecodedTransaction(transaction: PartiallyDecodedTransaction): NewTransaction {
+    const { messageBytes, signatures } = transaction;
+
+    /*
+    Relevant message structure is at the start:
+    - transaction version
+    - transaction header (includes number of signer accounts)
+    - static accounts (signers first)
+    */
+
+    let offset = 0;
+
+    // read version
+    offset = getTransactionVersionDecoder().read(messageBytes, offset)[1];
+
+    // read header
+    // first byte of header is numSignerAccounts
+    const numSignersResult = getU8Decoder().read(messageBytes, offset);
+    const numSigners = numSignersResult[0];
+    // there are two more bytes which we don't need to read
+    offset = numSignersResult[1] + 2;
+
+    // read signer addresses (these are first in the static accounts)
+    // compact-u16 array format, read the short-u16 length
+    const addressesLength = getShortU16Decoder().read(messageBytes, offset);
+    offset = addressesLength[1];
+
+    // by decoding this array with a fixed size,
+    // we don't try to read a length prefix and only read as many addresses as we need
+    const [signerAddresses] = getArrayDecoder(getAddressDecoder(), { size: numSigners }).read(messageBytes, offset);
+
+    // signer addresses and signatures must be the same length
+    // we encode an all-zero signature when the signature is missing
+    if (signerAddresses.length !== signatures.length) {
+        throw new SolanaError(SOLANA_ERROR__TRANSACTION__SIGNATURES_LENGTH_NUM_SIGNERS_MISMATCH, {
+            signaturesLength: signatures.length,
+            signerAddresses,
+            signerAddressesLength: signerAddresses.length,
+        });
+    }
+
+    // combine the signer addresses + signatures into the signatures map
+    const signaturesMap: SignaturesMap = {};
+    signerAddresses.forEach((address, index) => {
+        const signatureForAddress = signatures[index];
+        if (signatureForAddress.every(b => b === 0)) {
+            signaturesMap[address] = null;
+        } else {
+            signaturesMap[address] = signatureForAddress as SignatureBytes;
+        }
+    });
+
+    return {
+        messageBytes: messageBytes as TransactionMessageBytes,
+        signatures: signaturesMap,
```

Comment:
Signature map should be frozen at this stage (and we should have a test for it).

Reply:
Thanks! Test added: https://github.com/solana-labs/solana-web3.js/blob/4214293ce77332dff899468c01962d72a306fd34/packages/transactions/src/codecs/__tests__/transaction-codec-test.ts#L138